pkgs/build-support/rust/import-cargo-lock.nix

```diff
@@ -125,20 +125,31 @@ let
         };
       in runCommand "${pkg.name}-${pkg.version}" {} ''
         tree=${tree}
-        if grep --quiet '\[workspace\]' "$tree/Cargo.toml"; then
-          # If the target package is in a workspace, find the crate path
-          # using `cargo metadata`.
-          crateCargoTOML=$(${cargo}/bin/cargo metadata --format-version 1 --no-deps --manifest-path $tree/Cargo.toml | \
-            ${jq}/bin/jq -r '.packages[] | select(.name == "${pkg.name}") | .manifest_path')
 
+        # If the target package is in a workspace, or if it's the top-level
+        # crate, we should find the crate path using `cargo metadata`.
+        crateCargoTOML=$(${cargo}/bin/cargo metadata --format-version 1 --no-deps --manifest-path $tree/Cargo.toml | \
+          ${jq}/bin/jq -r '.packages[] | select(.name == "${pkg.name}") | .manifest_path')
+
+        # If the repository is not a workspace the package might be in a subdirectory.
+        if [[ -z $crateCargoTOML ]]; then
+          for manifest in $(find $tree -name "Cargo.toml"); do
```

Comment:
What is the search order in Cargo? I'd think that in such cases a breadth-first would be preferred (higher in the tree is better). Also, with a breadth first search, this could be combined with the earlier `cargo metadata` call.

tl;dr: do you know what the traversal order in Cargo is? We probably want to do the same.

Reply:
`cargo metadata` uses the [provided workspace member globs](https://github.com/rust-lang/cargo/blob/94ca096afbf25f670e76e07dca754fcfe27134be/src/cargo/core/workspace.rs#L1603-L1629) in the order they were given with the [globs expanding in alphabetical order](https://docs.rs/glob/0.3.0/glob/fn.glob.html).

since we are dealing with the case where the workspace members do not contain the desired crate, we can't viably use `cargo metadata` as an implementation guide. i don't have a strong opinion on the `find` order. both are predictable in knowing which crate will be found first in case there are multiple crates with the same name in the tree.